designs/credentialManagement/design.md

```diff
@@ -0,0 +1,115 @@
+more than one option is possible. Due to the nature of the IntellJ IDE being multiple windows in one JVM, each project
+window can have a different active credential selected.
+
+### Extension System
+
+We register all `ToolkitCredentialsProviderFactory` through the IntelliJ extension point system by creating a custom
+extension point bean (`CredentialProviderFactoryEP`) under the FQN `aws.toolkit.credentialProviderFactory` which has a 
+single `implementation` attribute. The extension system is only queried once, and the list of instances of
+`ToolkitCredentialsProviderFactory` must be immutable.
+
+Example of usage:
+```xml
+    <extensions defaultExtensionNs="aws.toolkit">
+        <credentialProviderFactory implementation="software.aws.toolkits.jetbrains.core.credentials.profiles.ProfileCredentialProviderFactory"/>
+    </extensions>
+```
+
+### Credential Validation
+
+In order to validate the credentials returned by the  `ToolkitCredentialsProvider` are valid, we make a call to 
+`sts::GetCallerInfo`.
+
+### Built-in Providers
+
+#### Shared Credentials File(s) (Profile file)
+
+Also known as credential profiles, this sources the credentials from the `~/.aws/config` and `~/.aws/credentials` file
+according to [CLI documentation][CliConfigDocs]. We try to comply as close as possible with the CLI behavior, but not
+all keys are supported in the toolkit. We ignore properties not related to credential management as well.
+
+Supported keys:
+* `aws_access_key_id`
+* `aws_secret_access_key`
+* `aws_session_token`
+* `source_profile`
+* `external_id`
+* `role_session_name`
+* `mfa_serial`
+* `credential_process`
+
+## User Experience Walkthrough
+
+### Status Bar
+
+We register a status bar widget in order to provide the user with an indicator of what credential is active in their 
+project. The status bar also acts as an entry point to switching by having a switcher in the context menu.
+
+![NoCredentialsStatusBar]
+
+*Image when no credentials are active*
+
+![DefaultCredentialsStatusBar]
+
+*Image when a profile named `default is active*
+
+### On Toolkit Start Run
+
+If the user has no active credentials from their last session, we should see if they have a profile named `default` 
+and select it for them if the credentials are valid.
+
+### On Switching Credentials
+
+When the user wishes to switch their active credentials, we first perform a check to see if the call is valid. See 
+[Credential Validation](#Credential-Validation).
+
+* If the check passes, we will mark the selected credentials active for the project window.
+* If the check fails, we will fall back to having no credentials specified and tell the user that the validation 
+failed.
+
+### Multi-Factor Authentication
+
+If the credential provider has MFA, we will need to prompt the user for their OTP. This works by blocking the 
+`resolveCredentials` call until a input dialog message prompt is filled in.
```

Comment:
Consider "This works by blocking the 
`resolveCredentials` call _in AwsCredentialProvider (plus link)_ ..."